homeassistant/components/go2rtc/__init__.py

```diff
@@ -203,91 +232,30 @@ async def _get_binary(hass: HomeAssistant) -> str | None:
-        session_id: str,
-        send_message: WebRTCSendMessage,
-    ) -> None:
-        """Handle the WebRTC offer and return the answer via the provided callback."""
-        self._sessions[session_id] = ws_client = Go2RtcWsClient(
-            self._session, self._url, source=camera.entity_id
-        )
+@callback
+def async_is_supported(stream_source: str | None) -> bool:
+    """Check if the stream source is supported by go2rtc."""
+    return (
+        stream_source is not None
+        and stream_source.partition(":")[0] in _SUPPORTED_STREAMS
+    )
 
-        if not (stream_source := await camera.stream_source()):
-            send_message(
-                WebRTCError("go2rtc_webrtc_offer_failed", "Camera has no stream source")
-            )
-            return
-
-        streams = await self._rest_client.streams.list()
-
-        if (stream := streams.get(camera.entity_id)) is None or not any(
-            stream_source == producer.url for producer in stream.producers
-        ):
-            await self._rest_client.streams.add(
-                camera.entity_id,
-                [
-                    stream_source,
-                    # We are setting any ffmpeg rtsp related logs to debug
-                    # Connection problems to the camera will be logged by the first stream
-                    # Therefore setting it to debug will not hide any important logs
-                    f"ffmpeg:{camera.entity_id}#audio=opus#query=log_level=debug",
-                ],
-            )
 
-        @callback
-        def on_messages(message: ReceiveMessages) -> None:
-            """Handle messages."""
-            value: WebRTCMessage
-            match message:
-                case WebRTCCandidate():
-                    value = HAWebRTCCandidate(RTCIceCandidateInit(message.candidate))
-                case WebRTCAnswer():
-                    value = HAWebRTCAnswer(message.sdp)
-                case WsError():
-                    value = WebRTCError("go2rtc_webrtc_offer_failed", message.error)
-
-            send_message(value)
-
-        ws_client.subscribe(on_messages)
-        config = camera.async_get_webrtc_client_configuration()
-        await ws_client.send(WebRTCOffer(offer_sdp, config.configuration.ice_servers))
-
-    async def async_on_webrtc_candidate(
-        self, session_id: str, candidate: RTCIceCandidateInit
-    ) -> None:
-        """Handle the WebRTC candidate."""
-
-        if ws_client := self._sessions.get(session_id):
-            await ws_client.send(WebRTCCandidate(candidate.candidate))
-        else:
-            _LOGGER.debug("Unknown session %s. Ignoring candidate", session_id)
-
-    @callback
-    def async_close_session(self, session_id: str) -> None:
-        """Close the session."""
-        ws_client = self._sessions.pop(session_id)
-        self._hass.async_create_task(ws_client.close())
+async def create_go2rtc_client(
+    camera: Camera, client_remove_fn: Callable[[], None]
```

Comment:
if you remove `camera` from `Go2RtcClient` as suggested by Martin, can we remove it here and replace it by `stream_source: str`? again to break (or at least reduce) the circular deps